### PR TITLE
Split service into a seperate SLS, manage dirs, more settings

### PR DESCRIPTION
**Pull request by cmclaughlin**
* Refactor for more control if/when service restarts
* Create log/data dirs if defined in pillar
* Populate defaults/sysconfig env vars from Pillar

### CODE REVIEW

**Comment by cmclaughlin** [comment]
@kpostrup here's another for you.

I also tested this one on Amazon Linux 2015.09 and Ubuntu 14.04

**Comment by jpic** [comment]
Thanks for adding more awesome distro support @cmclaughlin.

I need some support to understand this patch. The commit message mentions a service not starting, perhaps some logs could be added ? Also, what else does this patch to fix the service on other distros ? Perhaps you could contribute the patch that fixes the service too ? You could set up testing on multiple distros on travis-ci like by instanciating a bunch of full system containers and executing shell commands for [example](https://github.com/novafloss/ansible-role-boot/blob/c6b97132ff5a46f77ae8d12a7c9e4723cef94753/test.yml#L13-L16).

Then, could this patch fix the service forever ?

**Comment by kpostrup** [comment]
@jpic I don't think @cmclaughlin meant that the service needed fixing. This PR implements a split of the service into its own SLS making him able to selectively run the other states, except for the service.

I have no problem with this.

**Comment by jpic** [comment]
Do you mean that you prefer the user to:

- execute elasticsearch.install
- execute elasticsearch.config
- execute states to create paths
- execute elasticsearch.service

Rather than just:

- execute elasticsearch with abritrary paths

I would perhaps have wanted to compare both implementations before choosing one but perhaps I just don't get what's going on here, anyway, just trying to help, thanks for reading ! Keep up the great work !

:guitar:

**Comment by kpostrup** [comment]
@jpic I think you're right. It would probably make more sense if, instead of splitting the formula up as suggested. the state would check for the `path.data` and `path.data` directories in Pillar and depend on them to exist or create them, before starting the service. Is that what you mean?

**Comment by jpic** [comment]
@kpostrup yes I'm wondering if it would make sense to compare your solution with the one suggested in this PR before deciding what kind of fix do we want here. Do we want a curative solution for the issue described in the commit message or do we want to make it easier for users to implement paliative solutions ? I'm not a salt expert here so I can't say really, just suggesting the idea to compare.

**Comment by cmclaughlin** [comment]
I'll try to clarify my usecase. Here's how I'm using the formula:

	include:
	  # This only installs ElasticSearch
	  - elasticsearch.install

	  # This setups up the ES config file
	  - elasticsearch.config

	  # Here I create /mnt/elasticsearch/data (my path.data) and /mnt/elasticsearch/logs (my path.logs)
	  # These need to be created after the installation so they can be owned by the elasticsearch user,
	  # but before the service starts
	  - shopstyle-elasticsearch.dirs

	  # Here I setup /etc/default/elasticsearch... maybe we should do this in the elasticsearch-formula?
	  # Technically this starts the service, because of a watch_in on the service
	  - shopstyle-elasticsearch.config

	  # Manage the service
	  - elasticsearch.service
  
I hope this makes sense now... the service wasn't broken, I just wanted more control of the order.  Since Salt defaults to evaluating in the order in which resources are defined, I believe this allows me the control I need for my installation.

Regarding testing... I just tested this manually by launching instances on AWS.  I don't have any CI setup at the moment... it's something I'd like to do with Salt, but haven't had the time to try.

**Comment by cmclaughlin** [comment]
I just now noticed there were comments before my last one... I hadn't refreshed the page, so I hope what I said didn't seem out of context. 

I'd be happy to refactor the formula to create the path.data and path.logs directory instead of doing that in my own state.  Please let me know if you have a preference.

**Comment by jpic** [comment]
Hi Charles !

Go for it !

Rock'on !

:guitar:

File: elasticsearch/config.sls [hunk 1]
```diff
@@ -1,15 +1,11 @@
...
   file.managed:
     - name: /etc/elasticsearch/elasticsearch.yml
     - source: salt://elasticsearch/files/elasticsearch.yml
     - template: jinja
     - user: root
-    - watch_in:
-      - service: elasticsearch
     - require:
-      - pkg: elasticsearch
+      - sls: elasticsearch.install
```
Comment (iggy) [inline elasticsearch/config.sls]: Seems like an unnecessary change, but it shouldn't break anything.

File: elasticsearch/install.sls [hunk 2]
```diff
@@ -0,0 +1,10 @@
+include:
```
Comment (iggy) [inline elasticsearch/install.sls]: I think `pkg.sls` is a more standard name across the formulas

File: README.rst [hunk 3]
```diff
@@ -19,9 +19,14 @@ Available states
...
 
 ``elasticsearch.config``
 -----------
 
 Configures Elasticsearch.
+
+``elasticsearch.service``
+-----------
+
+Manages the Elasticsearch service.
```
Comment (iggy) [inline README.rst]: No documentation for the states that were split out (elasticsearch.install ?)

**Comment by cmclaughlin** [comment]
I just pushed some more commits.  Now I'm creating the dirs from Pillar data.  I still think it's useful to split out the service, so I left that change intact.

**Comment by cmclaughlin** [comment]
Added one more change - support for /etc/default/elasticsearch (Debian) and /etc/sysconfig/elasticsearch (RedHat).

I think I'm done... let me know what you think.

File: README.rst (hunk 3 above)
Comment (cmclaughlin) [inline README.rst]: Done!

**Comment by jpic** [comment]
It would have been nice to keep both features (split sls and add pillars) in separate commits but it's not important, just something reading the commit made me think about.

Thanks a lot for your contribution, keep up the great work !!

:guitar:

**Comment by cmclaughlin** [comment]
@kpostrup please let me know if there's anything left to do here before merging.

**Comment by jpic** [comment]
:guitar:
